pandas/core/indexes/base.py

```diff
@@ -2858,7 +2859,7 @@ def _intersection(self, other, sort=False):
             indexer = algos.unique1d(Index(rvals).get_indexer_non_unique(lvals)[0])
             indexer = indexer[indexer != -1]
 
-        result = other.take(indexer)._values
+        result = other.take(indexer).unique()._values
```

Comment:
can you add an assert that L2867 is *always* unique (e.g. we are guaranteed uniqu here L2862 (and I think safe_sort guaranteess this), but let's be explicit (and add a comment).

Reply:
you should check if this is_unique first. it will be computed but is likely faster than *always* doing this (and copies yet again).

Reply:
We may have a numpy array here, so we can not do result.is_unique. Is there a better way than using algos.unique on result and comparing shapes then?

Reply:
sure you can do (measure perf)
```
result = Index(other.take(indexer), copy=False)
if not result.is_unique:
 .....
```

Reply:
Sorry, did not get that. Did it now, is in #38154